kvcoord: deflake rangefeed tests

stevendanna opened the pull request:
This aims at fixing recent failures in

    TestRangeFeedMetricsManagement
    TestMuxRangeFeedCatchupScanQuotaReleased

A clear root cause was found for the former while the latter appears
to be related to slowdowns as a result of recreating catchup scan iterators
when running under the race-detector.

Fixes #156852
Fixes #157003
Fixes #157064

stevendanna committed:
rangefeed: recreate iterators less often under stress

Recreating the iterator substantially slows down some tests. For example,
locally I observed the following for TestMuxRangeCatchupScanQuotaReleased:

- No iterator recreation:       9s
- Recreation on every key:     90s
- Recreation on every 10 keys: 13s

Now, 90s is no where near the multi-minute timeout in the test. But, my laptop
is typically substantilly faster than CI.

Fixes #157003
Fixes #156852

Release note: none

stevendanna committed:
kvcoord: deflake TestRangeFeedMetricsManagement

This test failed to enable rangefeeds. One may ask why this ever worked at all.
My belief is that when the test was faster, the catch-up scans may have been
completing before all of the relevant ranges had explicitly set span
configuration. When no explicit span configuration is in place, we default to
assuming that the range has rangefeeds enabled via span configuration. When the
test is too slow, some ranges then observe that they _do not_ have rangefeeds
enabled via span configuration and fall back to the cluster setting.

The test assumed the rangefeed never encounter an error. Here we

1. Rework this test and some of the test helpers to be able to observe
   changefeed failures that occur before some expected event.

2. We enable rangefeeds for TestRangeFeedMetricsManagement.

Fixes #157064

Release note: None

stevendanna committed:
kvcoord: rename a few rangefeed tests

Epic: none
Release note: None

stevendanna committed:
kvcoord: speed up TestMuxRangeFeedCatchupScanQuotaReleased a bit

The number of errors here seemed somewhat arbitrary and cutting it down a bit
speeds the test up under race/stress.

Epic: none
Release note: none

tbg reviewed (approved):
> @tbg reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157259)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv and @sumeerbhola)

___
*[`pkg/kv/kvserver/rangefeed/catchup_scan.go` line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxYVnuA3wSSpmoFixV:-OdxYVnuA3wSSpmoFixW:ba3gmgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8721859475b6c3d7b6c7efa66bdcab2d8ba5b7/pkg/kv/kvserver/rangefeed/catchup_scan.go#L266)):*
> ```Go
> 				// Fibonacci hash https://probablydance.com/2018/06/16/fibonacci-hashing-the-optimization-that-the-world-forgot-or-a-better-alternative-to-integer-modulo/
> 				simpleHash := (11400714819323198485 * uint64(uintptr(unsafe.Pointer(&iter)))) >> 63
> 				forceRecreateUnderTest = simpleHash == 0
> ```

So this does it 50% of the time? That still seems like a lot.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go` line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxWJmDE3ncqsq3-d5r:-OdxWJmDE3ncqsq3-d5s:b-sm5j40) ([raw file](https://github.com/cockroachdb/cockroach/blob/227f59c83000d9e832ac26303945b307a0126364/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go#L185)):*
> ```Go
> func channelWaitWithTimeout(t *testing.T, ch chan struct{}, errCh chan error) {
> 	t.Helper()
> 	timeOut := testutils.SucceedsSoonDuration()
> ```

:drake-good: :thumbsup: 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go` line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxWY2SEtcBA-4USj-T:-OdxWY2SEtcBA-4USj-U:b-bcnhc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/227f59c83000d9e832ac26303945b307a0126364/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go#L221)):*
> ```Go
> 	startTime := ts.Clock().Now()
> 
> 	kvserver.RangefeedEnabled.Override(
> ```

It is funny. So when rangefeeds are off, we can still do catch up scans? Huh?


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
> @sumeerbhola reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157259)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv, @stevendanna, and @tbg)

___
*[`pkg/kv/kvserver/rangefeed/catchup_scan.go` line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxYVnuA3wSSpmoFixV:-Odxa5ezEjqvpQ8ExF3k:be7kw42) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8721859475b6c3d7b6c7efa66bdcab2d8ba5b7/pkg/kv/kvserver/rangefeed/catchup_scan.go#L266)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

So this does it 50% of the time? That still seems like a lot.
</blockquote></details>

Change this to 25% via `>> 62`?

Thanks for fixing this.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157259)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv and @tbg)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go` line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxWY2SEtcBA-4USj-T:-OdxaDZn49TbYhLij-Ip:b-h6fssx) ([raw file](https://github.com/cockroachdb/cockroach/blob/227f59c83000d9e832ac26303945b307a0126364/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go#L221)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It is funny. So when rangefeeds are off, we can still do catch up scans? Huh?
</blockquote></details>

Hrm. Actually, while I did observe this error and while adding this did resolve the issue, I think I mis-described the order of operations that must be happening because now that I re-read the code, we do seem to check this before we the catch up scan should be able to start.  I'll take a second look.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157259)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv and @tbg)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go` line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxWY2SEtcBA-4USj-T:-OdxeXk74Wq8fq1jZO9d:b-8znnxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/227f59c83000d9e832ac26303945b307a0126364/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go#L221)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Hrm. Actually, while I did observe this error and while adding this did resolve the issue, I think I mis-described the order of operations that must be happening because now that I re-read the code, we do seem to check this before we the catch up scan should be able to start.  I'll take a second look.
</blockquote></details>

I am not 100% sure, but one possibility is that some ranges are starting their catch up scan while others are hitting the error. One way that could happen is if the span configuration attach to our replica either hasn't been explicitly set et or because we are sharing a range with a span that has rangefeeds enabled via zone configuration:

```
func (r *Replica) isRangefeedEnabledRLocked() (ret bool) {
	if !r.mu.spanConfigExplicitlySet {
		return true
	}
	return r.mu.conf.RangefeedEnabled
}
```


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
Thanks for looking into this! Is `TestRangeFeedMetricsManagement` flaky now just because catch up scans are now slower with the re-opening iterators? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-:-OdxiuOz9GZkpGX8zgyP:b-90smz8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv and @tbg)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go` line 227 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxigGV-boADNorPS1d:-OdxigGW4e1N6qsAE5Cl:bgmywq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/227f59c83000d9e832ac26303945b307a0126364/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go#L227)):*
> ```Go
> 	// test cluster nodes.
> 	sqlDB.ExecMultiple(t,
> 		`ALTER DATABASE defaultdb CONFIGURE ZONE USING num_replicas = 1`,
> ```

Why doesn't cluster setting work? Is it because it needs rangefeed to set cluster settings or something else? 


<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/rangefeed/catchup_scan.go:
```diff
@@ -253,6 +256,16 @@ func (i *CatchUpSnapshot) CatchUpScan(
+			if util.RaceEnabled {
+				// Fibonacci hash https://probablydance.com/2018/06/16/fibonacci-hashing-the-optimization-that-the-world-forgot-or-a-better-alternative-to-integer-modulo/
+				simpleHash := (11400714819323198485 * uint64(uintptr(unsafe.Pointer(&iter)))) >> 63
+				forceRecreateUnderTest = simpleHash == 0
```
So the semantics here are: when we decide whether `&iter` recreated, this decision is sticky. But not guaranteed because the runtime can fiddle with the `uintptr`? Which is fine if it happens occasionally?

Contrasting to `rand % 2 == 0` <- that would recreate many times in all cases.

stevendanna replied:
I agree the previous semantics were maybe confusing. I tested this locally with just a integer modulo to do this every 10 keys rather than every key and it seems to be OK on these tests at least. If we keep seeing flakes we can perhaps move this to something simpler:

- A bool that forces us to recreate the iterator at least once under race the first time through this code and then just falls back to the normal behavior.
<!-- Reviewable comment -OdyS4AZ0qKYiFdJi3xP:b-7e0kai -->
<!-- Sent from Reviewable.io -->

pav-kv replied:
I'm good with either approach, just spelling out the semantics would be good. The pointer trick seems cool: if you end up with it, consider factoring out into a func with a comment.

Disclaimer: haven't looked at the code changes after the last time.

stevendanna reviewed (commented):
>  Is `TestRangeFeedMetricsManagement` flaky now just because catch up scans are now slower with the re-opening iterators?

I put my best theory in the commit message for my update to that test:

```
    This test failed to enable rangefeeds. One may ask why this ever worked at all.
    My belief is that when the test was faster, the catch-up scans may have been
    completing before all of the relevant ranges had explicitly set span
    configuration. When no explicit span configuration is in place, we default to
    assuming that the range has rangefeeds enabled via span configuration. When the
    test is too slow, some ranges then observe that they _do not_ have rangefeeds
    enabled via span configuration and fall back to the cluster setting.
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-:-OdyPxAbCeDQoB-K3lWN:b-1237k5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv, @sumeerbhola, @tbg, and @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/catchup_scan.go` line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxYVnuA3wSSpmoFixV:-OdyQBdGAF9ugOSYAz6W:bd7umkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8721859475b6c3d7b6c7efa66bdcab2d8ba5b7/pkg/kv/kvserver/rangefeed/catchup_scan.go#L266)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Change this to 25% via `>> 62`?

Thanks for fixing this.
</blockquote></details>

I changed this to every 10 keys.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go` line 227 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxigGV-boADNorPS1d:-OdySpSMBHvt8Dx__uPI:b-qj4dt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/227f59c83000d9e832ac26303945b307a0126364/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go#L227)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Why doesn't cluster setting work? Is it because it needs rangefeed to set cluster settings or something else?
</blockquote></details>

Unintended change. I started trying to get these tests to work with secondary tenants and then decided to punt that to another day.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157259)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv, @stevendanna, @tbg, and @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/catchup_scan.go` line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxYVnuA3wSSpmoFixV:-Odyd_Te-KHLlISoa29j:b-3b1ypr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8721859475b6c3d7b6c7efa66bdcab2d8ba5b7/pkg/kv/kvserver/rangefeed/catchup_scan.go#L266)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I changed this to every 10 keys.
</blockquote></details>

Some test have very few keys. And having those tests be randomly covered with immediate restarts was very helpful in finding various corner cases. The iterator hashing has that nice property.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> @tbg reviewed 2 of 2 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157259)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv and @wenyihu6)



<!-- Sent from Reviewable.io -->

stevendanna committed:
rangefeed: use FibHash in CatchUpScan

This brings back the use of FibHash to decide on whether to recreate an iterator
but now with a few more comments.

Epic: none
Release note: None

stevendanna committed:
util: add FibHash function

This adds a FibHash function that implements a "fibonacci hash." This is a
simple multiplicative hash that we use occasionally for testing Pebble and would
like to use for similar purposes in CockroachDB as well.

We also add Max and Min functions that take an iter.Seq.

Epic: none
Release note: None

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157259)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv, @sumeerbhola, @tbg, and @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/catchup_scan.go` line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-OdxYVnuA3wSSpmoFixV:-OdzYdigEhoZJT8d-RRW:bb9d1io) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b8721859475b6c3d7b6c7efa66bdcab2d8ba5b7/pkg/kv/kvserver/rangefeed/catchup_scan.go#L266)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Some test have very few keys. And having those tests be randomly covered with immediate restarts was very helpful in finding various corner cases. The iterator hashing has that nice property.
</blockquote></details>

OK. I've brought this back, but moved it into a helper method and added come comments per @pav-kv's suggestion.


<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/rangefeed/catchup_scan.go:
```diff
@@ -253,6 +255,22 @@ func (i *CatchUpSnapshot) CatchUpScan(
+			// proved a bit slow.
+			forceRecreateUnderTest := false
+			if util.RaceEnabled {
+				// Hash the pointer of our iterator into the range [0, 4] and only
```
nit: `[0,4)` or `[0,3]`

Optionally, remove one step that needs comprehending, and say that the probability is 1/4 (for a fixed `iter`).

pav-kv reviewed (approved):
Thank you!

pav-kv commented:
Please tag the fixed issues in the desc, for posterity etc.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @sumeerbhola reviewed 1 of 3 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157259#-:-Oe1XFgbE7ayqyWh06VZ:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @pav-kv, @stevendanna, @tbg, and @wenyihu6)



<!-- Sent from Reviewable.io -->

stevendanna commented:
TFTR!

bors r=pav-kv,tbg,sumeerbhola
